File: earnings-call-analyzer/web/src/ground.ts
// Grounding layer (differentiator #1). Mirrors eca/ground.py.
// Verifies every claim cites a real line, and every FINANCIAL figure (money/percent)
// in a claim appears on one of its cited lines. Incidental integers (counts, years) are
// not treated as claims to verify.

import type { Analysis, GroundingResult } from "./types.js";
import { SECTIONS } from "./types.js";
import { line, type Transcript } from "./ingest.js";

const MONEY_NUM = /\$\s?(\d[\d,]*(?:\.\d+)?)/gi;
const PCT_NUM = /(\d[\d,]*(?:\.\d+)?)\s?%/g;

function figures(text: string): Set<string> {
  const out = new Set<string>();
  for (const m of text.matchAll(MONEY_NUM)) out.add(m[1].replace(/,/g, ""));
  for (const m of text.matchAll(PCT_NUM)) out.add(m[1].replace(/,/g, ""));
  return out;
}

function claimText(section: string, item: any): string {
  if (section === "key_metrics") return `${item.metric ?? ""} ${item.value ?? ""}`;
  return item.text ?? "";
}

export function checkGrounding(analysis: Analysis, t: Transcript): GroundingResult {
  const r: GroundingResult = {
    total_claims: 0, cited_claims: 0, numeric_claims: 0, numerically_verified: 0,
    coverage: 1, numeric_accuracy: 1, issues: [],
  };
  const nLines = t.lines.length;

  for (const section of SECTIONS) {
    const items = (analysis as any)[section] as any[];
    for (const item of items ?? []) {
      r.total_claims += 1;
      const text = claimText(section, item);
      const cites: number[] = Array.isArray(item.citations) ? item.citations.filter((c: unknown) => Number.isInteger(c)) : [];
      const valid = cites.filter((c) => c >= 1 && c <= nLines);
      if (valid.length === 0) {
        r.issues.push(`[${section}] uncited or out-of-range: ${JSON.stringify(text.slice(0, 80))}`);
        continue;
      }
      r.cited_claims += 1;
      const claimFigures = figures(text);
      if (claimFigures.size > 0) {
        r.numeric_claims += 1;
        const cited = valid.map((c) => line(t, c) ?? "").join(" ");
        const sourceFigures = figures(cited);
        const overlap = [...claimFigures].some((f) => sourceFigures.has(f));
        if (overlap) r.numerically_verified += 1;
        else r.issues.push(`[${section}] figure(s) ${[...claimFigures].sort().join(", ")} not on cited line(s) ${valid.join(", ")}: ${JSON.stringify(text.slice(0, 80))}`);
      }
    }
  }
  r.coverage = r.total_claims ? r.cited_claims / r.total_claims : 1;
  r.numeric_accuracy = r.numeric_claims ? r.numerically_verified / r.numeric_claims : 1;
  return r;
}

File: earnings-call-analyzer/eca/ground.py
"""Grounding layer (differentiator #1).

Hallucinated figures are the adoption blocker on investment desks. So every claim in
the analysis must cite a real transcript line, and every number in a claim must actually
appear on one of its cited lines. This module verifies both and reports coverage.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from . import prompts
from .ingest import Transcript

# Verification targets FINANCIAL figures — money amounts and percentages — not every
# integer. Lexicon counts, years, and line references are not claims to verify.
_MONEY_NUM = re.compile(r"\$\s?(\d[\d,]*(?:\.\d+)?)", re.I)
_PCT_NUM = re.compile(r"(\d[\d,]*(?:\.\d+)?)\s?%")


@dataclass
class GroundingResult:
    total_claims: int = 0
    cited_claims: int = 0            # claims with >=1 resolvable citation
    numerically_verified: int = 0    # numeric claims whose number appears on a cited line
    numeric_claims: int = 0
    issues: list[str] = field(default_factory=list)

    @property
    def coverage(self) -> float:
        return (self.cited_claims / self.total_claims) if self.total_claims else 1.0

    @property
    def numeric_accuracy(self) -> float:
        return (self.numerically_verified / self.numeric_claims) if self.numeric_claims else 1.0


def _claim_text(section: str, item: dict) -> str:
    if section == "key_metrics":
        return f"{item.get('metric', '')} {item.get('value', '')}"
    return item.get("text", "")


def _nums(text: str) -> set[str]:
    """Financial figures in `text`, normalized (commas stripped)."""
    figures = _MONEY_NUM.findall(text) + _PCT_NUM.findall(text)
    return {f.replace(",", "") for f in figures}


def check_grounding(analysis: dict, transcript: Transcript) -> GroundingResult:
    result = GroundingResult()
    n_lines = len(transcript.lines)

    for section in prompts.SECTIONS:
        for item in analysis.get(section, []):
            result.total_claims += 1
            text = _claim_text(section, item)
            cites = [c for c in item.get("citations", []) if isinstance(c, int)]

            valid_cites = [c for c in cites if 1 <= c <= n_lines]
            if not valid_cites:
                result.issues.append(f"[{section}] uncited or out-of-range: {text[:80]!r}")
                continue
            result.cited_claims += 1

            claim_nums = _nums(text)
            if claim_nums:
                result.numeric_claims += 1
                cited_text = " ".join(transcript.line(c) or "" for c in valid_cites)
                source_nums = _nums(cited_text)
                if claim_nums & source_nums:
                    result.numerically_verified += 1
                else:
                    result.issues.append(
                        f"[{section}] number(s) {sorted(claim_nums)} not found on cited "
                        f"line(s) {valid_cites}: {text[:80]!r}"
                    )
    return result
